README.md
# RabbitMQ

A very simple and small services to simplify process to connect and call remote procedures (async way).

-   Project licensed under: GPLv3
-   Site Documentation: [Homepage](https://nsfilho.github.io/rabbitmq/index.html)
-   Repository: [GitHub](https://github.com/nsfilho/rabbitmq.git)

## Environment Variables

This services use some environment variables to pre-adjust some things, like:

-   `RABBITMQ_URL`: url to connect to amqp server. Default: `amqp://admin:password@localhost:5672`;
-   `RABBITMQ_RETRIES_INTERVAL`: interval to retry connection in milliseconds. Default: `1000`;
-   `RABBITMQ_ROUTINGKEY_PREFIX`: prefix to routing key to identify this as a sender. Default: `api`;
-   `RABBITMQ_DEBUG_CONSOLE`: show in console messages about rabbitmq status. Default: `false`;
-   `RABBITMQ_ENCONDING_CHARSET`: permits you to change buffer charset to encoding your messages. Default: `utf-8`
-   `RABBITMQ_INTERVAL_CONNECTION_CHECK`: interval to check (if multiples getConnections are called simultaneously) if it is connected. Default: `100`

## Running RabbitMQ

For use this library you will need to run a RabbitMQ. A sample docker-compose:

```yml
version: '2.0'

services:
    queue:
        image: rabbitmq:3-management
        ports:
            - 5672:5672
            - 15672:15672
        environment:
            - RABBITMQ_DEFAULT_USER=admin
            - RABBITMQ_DEFAULT_PASS=password
        volumes:
            - rabbitmq:/var/lib/rabbitmq

volumes:
    rabbitmq:
```

If you would like to use a docker swarm (stack) version, you can see in your sample folder.

## Example

You will found other samples in: [Samples Folder](https://github.com/nsfilho/rabbitmq/tree/master/sample).

```ts
/* eslint-disable no-console */
import { listenProcedureCall, remoteProcedureCall, getConnection, assertExchange } from '../src';

interface ExamplePayloadRequest {
    name: string;
    date: string;
    waitTimeToReturn: number;
}

interface ExamplePayloadReturn {
    timesCalled: number;
}

const delay = (timer: number) =>
    new Promise((resolve) => {
        setTimeout(() => {
            resolve();
        }, timer);
    });

const calledTimes = {
    x: 0,
    next: function next() {
        this.x += 1;
        return this.x;
    },
};

/**
 * Listen a procedure
 *
 * This will permit you to decoupling your code in a very nice way.
 * You can do many triggers or answers to remote requests
 *
 */
listenProcedureCall<ExamplePayloadRequest, ExamplePayloadReturn>({
    // Function name
    queue: 'exampleRemoteFunction',

    // Piece of code to execute 🚀
    callback: async ({ payload, stop }) => {
        const { waitTimeToReturn, name, date } = payload;
        const timesCalled = calledTimes.next();

        /** Logging information for you see the process */
        console.log(`Function called, with: ${name} at ${date}`);
        await delay(waitTimeToReturn);
        console.log(`Returning: ${timesCalled}`);

        /** stop listen more */
        stop();

        /** return the answer for remote caller */
        return {
            timesCalled,
        };
    },
});

const execute = async () => {
    // Call a remote procedure (so simple 🥰)
    const remoteReturn = await remoteProcedureCall({
        exchange: 'exampleRemoteFunction',
        payload: {
            name: 'inside function',
            date: new Date().toISOString(),
            waitTimeToReturn: 3000,
        },
        routingKey: 'default',
        assertReturnQueue: true,
        exclusiveReturnChannel: false,
        ignoreReturn: false,
    });

    // Log return for you see the fully process.
    console.log('Remote Return:', remoteReturn);
    await connection.close();
};

execute();
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.1.8](https://github.com/nsfilho/rabbitmq/compare/v0.1.7...v0.1.8) (2020-10-02)


### Features

* **channels:** improving channels systems ([211febd](https://github.com/nsfilho/rabbitmq/commit/211febd9ef2837885054a763c9b4d545f1aada62))

### [0.1.7](https://github.com/nsfilho/rabbitmq/compare/v0.1.6...v0.1.7) (2020-09-10)


### Features

* **connection:** guarantee only one connection ([8f1f73c](https://github.com/nsfilho/rabbitmq/commit/8f1f73c33255383f59225d2f2553d1d84e113ebd))

### [0.1.6](https://github.com/nsfilho/rabbitmq/compare/v0.1.5...v0.1.6) (2020-09-09)


### Bug Fixes

* changing RABBITMQ_DEBUG_CONSOLE ([9b0067b](https://github.com/nsfilho/rabbitmq/commit/9b0067bd4ea96b348c6699e436f5243638cb97ee))

### [0.1.5](https://github.com/nsfilho/rabbitmq/compare/v0.1.4...v0.1.5) (2020-09-09)


### Bug Fixes

* **rpc:** fixing return with exclusive id ([5ba90c1](https://github.com/nsfilho/rabbitmq/commit/5ba90c13546c1d1d4e75a984f9178dd9b3cdffc7))

### [0.1.4](https://github.com/nsfilho/rabbitmq/compare/v0.1.3...v0.1.4) (2020-09-06)


### Bug Fixes

* **docs:** fixing documentation site and urls ([79550a0](https://github.com/nsfilho/rabbitmq/commit/79550a0d08df903e985da249feff0c15a8646a8a))
* **docs:** reducing files to package ([9482096](https://github.com/nsfilho/rabbitmq/commit/948209620bb309a205da64482d60fe247693457e))

### [0.1.3](https://github.com/nsfilho/rabbitmq/compare/v0.1.2...v0.1.3) (2020-09-06)


### Features

* **rpc:** a lot of new features and parameters to make more flexible ([bf2222a](https://github.com/nsfilho/rabbitmq/commit/bf2222a69527aca14b32869f1b0b4c334a12b52f))
* **sample:** adding new samples ([00c8108](https://github.com/nsfilho/rabbitmq/commit/00c81080abefed92510df4fd659831ebf738f218))

### [0.1.2](https://github.com/nsfilho/rabbitmq/compare/v0.1.1...v0.1.2) (2020-09-05)


### Bug Fixes

* listening function ([bbc1b9b](https://github.com/nsfilho/rabbitmq/commit/bbc1b9be135447c3ca9eaf6ac578a67137d00047))

### 0.1.1 (2020-09-05)


### Features

* initial version ([7566e16](https://github.com/nsfilho/rabbitmq/commit/7566e16631a77dc2faaa9ce30ee4ac856c8459d4))


sample/loopingClient.ts
/* eslint-disable no-console */
import { remoteProcedureCall, getConnection, assertExchange, assertQueue } from '../src';

interface ExamplePayloadRequest {
    name: string;
    date: string;
    waitTimeToReturn: number;
}

interface ExamplePayloadReturn {
    timesCalled: number;
}

const execute = async () => {
    // --------------------------------------------------------------------------------------------------
    // bind the Exchange to Queue (via code, for test purpose only -- not need in production environment)
    // Generally DevOps will do the properly assignments in production environment.
    const connection = await getConnection();
    const channel = await connection.createChannel();
    await assertExchange({
        name: 'exampleRemoteFunction',
        type: 'fanout',
        advanced: {
            autoDelete: false,
            durable: true,
        },
    });
    await assertQueue({
        name: 'exampleRemoteFunction.return',
        advanced: {
            autoDelete: false,
            durable: true,
        },
    });
    await channel.bindQueue('exampleRemoteFunction', 'exampleRemoteFunction', 'default');
    await channel.close();
    // --------------------------------------------------------------------------------------------------

    setInterval(async () => {
        // Call a remote procedure (so simple 🥰)
        const remoteReturn = await remoteProcedureCall({
            exchange: 'exampleRemoteFunction',
            payload: {
                name: 'inside function',
                date: new Date().toISOString(),
                waitTimeToReturn: 3000,
            },
            routingKey: 'default',
        });

        // Log return for you see the fully process.
        console.log('Remote Return:', remoteReturn);
    }, 1000);
};

execute();
